zkevm-circuits/src/evm_circuit/execution/mul_div_mod_shl_shr.rs

```diff
@@ -0,0 +1,351 @@
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        step::ExecutionState,
+        util::{
+            self,
+            common_gadget::SameContextGadget,
+            constraint_builder::{ConstraintBuilder, StepStateTransition, Transition::Delta},
+            math_gadget::{IsZeroGadget, LtWordGadget, MulAddWordsGadget},
+            sum, CachedRegion, Cell,
+        },
+        witness::{Block, Call, ExecStep, Transaction},
+    },
+    util::Expr,
+};
+use bus_mapping::evm::OpcodeId;
+use eth_types::{Field, ToLittleEndian, U256};
+use halo2_proofs::plonk::{Error, Expression};
+
+/// MulDivModShlShrGadget verifies opcode MUL, DIV, MOD, SHL and SHR.
+/// For MUL, verify pop1 * pop2 == push;
+/// For DIV, verify pop1 / pop2 == push;
+/// For MOD, verify pop1 % pop2 == push;
+/// For SHL, verify pop1 * (2^pop2) == push;
+/// For SHR, verify pop1 / (2^pop2) = push;
+/// when pop1, pop2, push are 256-bit words.
+#[derive(Clone, Debug)]
+pub(crate) struct MulDivModShlShrGadget<F> {
+    same_context: SameContextGadget<F>,
+    /// Words quotient, divisor, remainder, dividend and
+    /// pop1 (used for shift value of SHL and SHR)
+    pub words: [util::Word<F>; 5],
```

Comment:
I think it'd be better to just have 
```
quotient: Word<F>,
divisor: Word<F>,
...
```
Since this is how they end up being used. There's a never a need to use them are an array for 5 Word's.